porcupine/plugins/anchors.py

```diff
@@ -0,0 +1,173 @@
+from __future__ import annotations
+import tkinter
+import time
+from typing import List
+from porcupine import tabs, get_tab_manager
+from porcupine.plugins.linenumbers import LineNumbers
+
+# Dependent on code from linenumbers.py
+setup_after = ["linenumbers"]
+
+
+class AnchorManager:
+    def __init__(self, tab_textwidget: tkinter.Text, linenumbers: tkinter.Canvas) -> None:
+        self.tab_textwidget = tab_textwidget
+        self.linenumbers = linenumbers
+
+        self.anchor_symbol = "¶"
+
+        # tkinter have default marks such as "insert", "current", "tkinter::anchor1"
+        self.custom_anchor_prefix = "anchor_"
+
+        linenumbers.bind("<<Updated>>", self.update_linenumbers, add=True)
+
+    def _get_anchors(self) -> List:
+        return [
+            anchor
+            for anchor in self.tab_textwidget.mark_names()
+            if anchor.startswith(self.custom_anchor_prefix)
+        ]
+
+    def _get_cursor_index(self):
+        return self.tab_textwidget.index("insert linestart")
+
+    def toggle_on_off(self, event: tkinter.Event[tkinter.Misc]) -> None:
+
+        self.prevent_duplicate_anchors()
+
+        cursor_index = self._get_cursor_index()
+        for anchor in self.tab_textwidget.mark_names():
+            if (
+                anchor.startswith(self.custom_anchor_prefix)
+                and self.tab_textwidget.index(anchor) == cursor_index
+            ):
+                self.tab_textwidget.mark_unset(anchor)
+                self.linenumbers.do_update()
+                return
+        self.tab_textwidget.mark_set(
+            self.custom_anchor_prefix + str(time.time()), "insert linestart"
+        )
+
+        self.linenumbers.do_update()
+
+    def jump_to_next(self, event: tkinter.Event[tkinter.Misc]) -> None:
+        cursor_row = self._get_cursor_index().split(".")[0]
+        anchor_list = self._get_anchors()
+        anchor_rows = reversed(
+            sorted(
+                [
+                    int(self.tab_textwidget.index(anchorpoint).split(".")[0])
+                    for anchorpoint in anchor_list
+                ]
+            )
+        )
+
+        next_anchor_row = 0
+        for anchor_row in anchor_rows:
+            if anchor_row > int(cursor_row):
+                next_anchor_row = anchor_row
+            else:
+                break
+
+        if not next_anchor_row:
+            return
+        else:
+            self.tab_textwidget.mark_set("insert", f"{str(next_anchor_row)}.0")
+
+        # If cursor is below last row
+        if int(self._get_cursor_index().split(".")[0]) > int(
+            self.tab_textwidget.index(f"@0,{self.tab_textwidget.winfo_height()}").split(".")[0]
+        ):
+            self.tab_textwidget.see("insert")
```

Comment:
It's fine to call `.see("insert")` unconditionally. It will do nothing if the cursor is already visible.

```suggestion
        self.tab_textwidget.see("insert")
```

Reply:
But I was thinking that if two anchors are visible and you jump, .see will make the "camera" jump.. Whereas I'd like it to not move if both anchors are visible.

Reply:
It shouldn't jump. From manual page:

> If index is already visible then              the command does nothing. If index is a short  distance  out  of              view,  the  command  adjusts  the view just enough to make index              visible at the edge of the window.  If index is far out of view,              then the command centers index in the window.

Reply:
Somehow the suggestion didn't click applied. Maybe you accidentally clicked "Resolve conversation" instead of "Commit suggestion"?